### ISSUE TITLE
Unexpected token error from scriptmanager.js

### ISSUE THREAD

**Issue by rmoedt** [hyperledger/composer#1611]
## Context
We have a Node.js application running inside a docker container that uses the composer-client to connect to a composer business network.
Randomly calls from the application to the composer network fail. Sometimes it fails once, sometimes it fails many times in a row. 
It happens when doing new transactions and also on calls to retrieve assets.

## Actual Behavior
scriptmanager.js:67:16 throws the following error: Unexpected token (77:0)
Character 77:0 is the last character in one of the javascript files inside the bna. We have multiple js files and the every time the error occurs it is on the last character of one of the files. The same happens when we put all the code in a single file. It doesn't matter if we have a extra white line at the end of a file or not.

The error in the console of the node.js app
[error.txt](https://github.com/hyperledger/composer/files/1160001/error.txt)

The last lines of the chaincode logs
[chaincode logs.txt](https://github.com/hyperledger/composer/files/1160004/chaincode.logs.txt)

Line 42 and 45 had the entire base64 encoded bna file. When I decode it and viewed the contents I could see all the js files and they look fine, no unexpected tokens at the end of any file.

## Possible Fix
Tried to exclude various things:
Ran the application locally outside docker: still the same issue
Made the fabric network simpler with just one peer, one channel, one deployed bna: still the same issue
Tried to replicate the issue by using the composer rest server, haven't seen the same issue there so far

## Steps to Reproduce
Create a fabric v1 network
Deploy a bna
Make calls to the network using the composer-cli

## Your Environment
macOS Sierra
Composer 0.9.2
Fabric x86_64-1.0.0-beta
Node 6.11.1
Docker Version 17.06.0-ce-mac18 (18433)

**Comment by bestbeforetoday** [comment]
[chaincode logs.txt](https://github.com/hyperledger/composer/files/1160004/chaincode.logs.txt) looks to be showing a successful invocation of chaincode. If something is failing then that log isn't showing it.

[error.txt](https://github.com/hyperledger/composer/files/1160001/error.txt) shows a failure of the Acorn JavaScript parser to parse some file that has been passed to it by the composer code. I think this is happening at the point the .bna file is being unpacked, but the stack trace doesn't go all the way to the top to show me. There is no record of the JS file it is failing on to help pin it down either.

It isn't immediately obvious to me how the errors supplied match up with the failures you are seeing making calls to your deployed business network.

Could you provide your .bna file or a cut-down variant that will allow me to reproduce the problem?

**Comment by davidkel** [comment]
Interesting here is a failure from our own system tests
```
  Transaction system tests
    1) "before all" hook
  Transaction (participant specific) system tests
    2) "before all" hook
  Transaction (query specific) system tests
    3) "before all" hook

1) Transaction system tests "before all" hook:
     SyntaxError: Unexpected token (26:0)
      at Parser.pp$4.raise (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:2488:13)
      at Parser.pp.unexpected (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:623:8)
      at Parser.pp$3.parseExprAtom (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:2011:10)
      at Parser.pp$3.parseExprSubscripts (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1872:19)
      at Parser.pp$3.parseMaybeUnary (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1849:17)
      at Parser.pp$3.parseExprOps (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1791:19)
      at Parser.pp$3.parseMaybeConditional (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1774:19)
      at Parser.pp$3.parseMaybeAssign (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1750:19)
      at Parser.pp$3.parseExpression (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1722:19)
      at Parser.pp$1.parseStatement (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:777:45)
      at Parser.pp$1.parseTopLevel (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:672:23)
      at Parser.parse (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:529:15)
      at Object.parse (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:3378:37)
      at new JavaScriptParser (/home/travis/build/hyperledger/composer/packages/composer-common/lib/codegen/javascriptparser.js:58:25)
      at new Script (/home/travis/build/hyperledger/composer/packages/composer-common/lib/introspect/script.js:48:24)
      at ScriptManager.createScript (/home/travis/build/hyperledger/composer/packages/composer-common/lib/scriptmanager.js:61:16)
      at scriptFiles.forEach (systest/transactions.js:47:51)
      at Array.forEach (native)
      at Context.<anonymous> (systest/transactions.js:45:21)
  2) Transaction (participant specific) system tests "before all" hook:
     SyntaxError: Unexpected token (361:0)
      at Parser.pp$4.raise (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:2488:13)
      at Parser.pp.unexpected (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:623:8)
      at Parser.pp$3.parseExprAtom (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:2011:10)
      at Parser.pp$3.parseExprSubscripts (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1872:19)
      at Parser.pp$3.parseMaybeUnary (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1849:17)
      at Parser.pp$3.parseExprOps (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1791:19)
      at Parser.pp$3.parseMaybeConditional (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1774:19)
      at Parser.pp$3.parseMaybeAssign (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1750:19)
      at Parser.pp$3.parseExpression (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1722:19)
      at Parser.pp$1.parseStatement (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:777:45)
      at Parser.pp$1.parseTopLevel (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:672:23)
      at Parser.parse (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:529:15)
      at Object.parse (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:3378:37)
      at new JavaScriptParser (/home/travis/build/hyperledger/composer/packages/composer-common/lib/codegen/javascriptparser.js:58:25)
      at new Script (/home/travis/build/hyperledger/composer/packages/composer-common/lib/introspect/script.js:48:24)
      at ScriptManager.createScript (/home/travis/build/hyperledger/composer/packages/composer-common/lib/scriptmanager.js:61:16)
      at scriptFiles.forEach (systest/transactions.participants.js:47:51)
      at Array.forEach (native)
      at Context.<anonymous> (systest/transactions.participants.js:45:21)
  3) Transaction (query specific) system tests "before all" hook:
     SyntaxError: Unexpected token (39:0)
      at Parser.pp$4.raise (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:2488:13)
      at Parser.pp.unexpected (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:623:8)
      at Parser.pp$3.parseExprAtom (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:2011:10)
      at Parser.pp$3.parseExprSubscripts (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1872:19)
      at Parser.pp$3.parseMaybeUnary (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1849:17)
      at Parser.pp$3.parseExprOps (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1791:19)
      at Parser.pp$3.parseMaybeConditional (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1774:19)
      at Parser.pp$3.parseMaybeAssign (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1750:19)
      at Parser.pp$3.parseExpression (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:1722:19)
      at Parser.pp$1.parseStatement (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:777:45)
      at Parser.pp$1.parseTopLevel (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:672:23)
      at Parser.parse (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:529:15)
      at Object.parse (/home/travis/build/hyperledger/composer/node_modules/acorn/dist/acorn.js:3378:37)
      at new JavaScriptParser (/home/travis/build/hyperledger/composer/packages/composer-common/lib/codegen/javascriptparser.js:58:25)
      at new Script (/home/travis/build/hyperledger/composer/packages/composer-common/lib/introspect/script.js:48:24)
      at ScriptManager.createScript (/home/travis/build/hyperledger/composer/packages/composer-common/lib/scriptmanager.js:61:16)
      at scriptFiles.forEach (systest/transactions.queries.js:119:51)
      at Array.forEach (native)
      at Context.<anonymous> (systest/transactions.queries.js:117:21)
```
And the build was triggered by a change to our docs, not to any code which should have had no influence on this and must be intermittent as other builds pass without any problem.

**Comment by davidkel** [comment]
While the build job still exists, see
https://travis-ci.org/hyperledger/composer/jobs/260697090

**Comment by davidkel** [comment]
Looking at the numbers. 26, 361, 39. These are all blank lines at the end of the files 
transactions.utilities.js, transactions.participants.js, transactions.queries.js

What is strange is that it is intermittent (non-deterministic), all most all of the time these files don't have a problem just very rarely they do.

**Comment by davidkel** [comment]
Looking at the numbers. 26, 361, 39. These are all blank lines at the end of the files 
transactions.utility.js, transactions.participants.js, transactions.queries.js

What is strange is that it is intermittent (non-deterministic), all most all of the time these files don't have a problem just very rarely they do.

**Comment by woutersm** [comment]
I'm working on the same team as Rik. And we are experiencing the same indeed. Do you still need something to reproduce the problem? I suspect (& hope) the failing tests are enough

**Comment by davidkel** [comment]
recreation is a problem here because these system tests don't fail, and these failed with this problem out of the blue and have not failed again since. 
How easily recreatable is it for you ? One thing we can consider is making sure that the JS files have absolutely no whitespace or hidden characters at the end of anything JS specific

**Comment by woutersm** [comment]
I've removed all the whitespace at the end of the different .js files. Makes no difference.

**Comment by davidkel** [comment]
@woutersm We don't have a recreate unfortunately, can you send us your failing example. Although our system tests failed once, they haven't failed since and could have been a problem with the file system

**Comment by davidkel** [comment]
Could you attach the chaincode container logs (assuming they are running with Debug on, which they would be if using the fabric-dev-server we provide) when a failure occurs. 
I am also assuming you are using couchdb if you are our dev server. I am wondering if there is a problem with the fabric layer returning information about the business network with spurious end characters sometimes such as a x'00' character at the end

**Comment by bestbeforetoday** [comment]
Two possible actions that _might_ help mitigate or resolve this behaviour are:
- Update to the latest version of the Acorn JavaScript parser.
- Switch from Duktape to Node.js as the JavaScript runtime in the chaincode container (#61)

**Comment by mbwhite** [comment]
FYI _ I've (unintentionally) reproduced this in my systest at the moment. Looking at the files in a hex editor to see if I can see anything (which I currently can't btw)

**Comment by mbwhite** [comment]
```SyntaxError: Failed to parse identities.js: Unexpected TokenError
      at Parser.pp.unexpected (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:638:6)
      at Parser.pp$3.parseExprAtom (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:2097:10)
      at Parser.pp$3.parseExprSubscripts (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:1957:19)
      at Parser.pp$3.parseMaybeUnary (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:1934:17)
      at Parser.pp$3.parseExprOps (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:1876:19)
      at Parser.pp$3.parseMaybeConditional (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:1859:19)
      at Parser.pp$3.parseMaybeAssign (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:1834:19)
      at Parser.pp$3.parseExpression (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:1805:19)
      at Parser.pp$1.parseStatement (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:799:45)
      at Parser.pp$1.parseTopLevel (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:692:23)
      at Parser.parse (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:543:15)
      at Object.parse (/home/matthew/github/composer/node_modules/acorn/dist/acorn.js:3672:37)
      at new JavaScriptParser (/home/matthew/github/composer/packages/composer-common/lib/codegen/javascriptparser.js:66:25)
      at new Script (/home/matthew/github/composer/packages/composer-common/lib/introspect/script.js:67:22)
      at ScriptManager.createScript (/home/matthew/github/composer/packages/composer-common/lib/scriptmanager.js:61:16)
      at scriptFiles.forEach (systest/identities.js:53:51)
      at Array.forEach (native)
      at Context.<anonymous> (systest/identities.js:51:21)
  ```
 
```TokenType {
  label: 'eof',
  keyword: undefined,
  beforeExpr: false,
  startsExpr: false,
  isLoop: false,
  isAssign: false,
  prefix: false,
  postfix: false,
  binop: null,
  updateContext: null }```

**Comment by mbwhite** [comment]
We have implemented a solution to this; though we haven't been able to determine accurately what the core issue is.

Within the acron parser, there is a main loop of code. 

```
                    while (this.type !== types.eof) {
                        let stmt = this$1.parseStatement(true, true, exports);
                        node.body.push(stmt);
                    }
```

What we have discovered is that when the end of the file reached, the token `types.eof` is set to `this.type`. However the loop will *sometimes* continue as the equality check is not returning the expected value. 

Checking the number times, type is called, duplicate copies of acorn in the codebase etc. has not shown any information indicating the real cause of the problem.

Theory indicates that Acorn is correct - however the practical side  is this equality check is failing.

Implemented a work around.. so that we change the equality of the main function.
